frontend/src/features/logic/utils/getVisibleFieldIds.ts

```diff
@@ -0,0 +1,205 @@
+          // then the condition must be satisfied if the current value is the special input value AND
+          // if the othersInput subfield has a value.
+          const satisfiesOthers =
+            currentValue.value === RADIO_OTHERS_INPUT_VALUE &&
+            !!currentValue.othersInput
+          if (satisfiesOthers) return true
+        }
+        return condValuesArray.includes(currentValue.value)
+      }
+      return condValuesArray.includes(currentValue)
+    }
+    case LogicConditionState.Equal: {
+      // Checkbox cannot have equal logic.
+      // Condition values cannot be an array for equality logic.
+      if (
+        Array.isArray(condition.value) ||
+        validateCheckboxInput(currentValue)
+      ) {
+        return false
+      }
+      if (validateRadioInput(currentValue)) {
+        if (condition.value === 'Others') {
+          // If the condition value is 'Others',
+          // then the condition must be satisfied if the current value is the special input value AND
+          // if the othersInput subfield has a value.
+          return (
+            currentValue.value === RADIO_OTHERS_INPUT_VALUE &&
+            !!currentValue.othersInput
+          )
+        }
+        return condition.value === currentValue.value
+      }
+      return String(condition.value) === currentValue
+    }
+  }
+}
+
+/**
+ * Checks if an array of conditions is satisfied.
+ * @param formInputs the inputs to the form to retrieve logic units for.
+ * @param logicUnit the logic units to check.
+ * @param visibleFieldMap the map with visible field ids as keys, which is used to ensure that conditions are visible
+ * @returns true if all the conditions are satisfied, false otherwise
+ */
+const isLogicUnitSatisfied = (
+  formInputs: Record<FormFieldDto['_id'], unknown>,
+  logicUnit: FormCondition[],
+  visibleFieldMap: FieldIdToType,
+): boolean => {
+  return logicUnit.every((condition) => {
+    const inputFieldType = visibleFieldMap[condition.field]
+    // If the field is not visible, then the field type will not be in the map.
+    if (!inputFieldType) return false
+    const input = formInputs[condition.field]
+    return (
+      input !== undefined &&
+      isConditionFulfilled(input, condition, inputFieldType)
+    )
+  })
+}
+
+/**
+ * Gets the IDs of visible fields in a form according to its responses.
+ * This function loops through all the form fields until the set of visible
+ * fields no longer changes. The first loop adds all the fields with no
+ * conditions attached, the second adds fields which are made visible due to fields added in the previous loop, and so on.
+ * @param formInputs the field responses to retrieve logic units for.
+ * @param formProps the form fields and form logics to determine visible field ids.
+ * @returns a set of IDs of visible fields in the submission
+ */
+export const getVisibleFieldIds = (
+  formInputs: Record<FormFieldDto['_id'], unknown>,
```

Comment:
would it not be possible to reuse the existing logic module, perhaps with an adapter layer?

Reply:
I'll see what I can do